perf(semantic): do not record ast nodes for cfg if cfg disabled

overlookmotel opened the pull request:
Control flow graph builder records AST node IDs in a temp structure `ast_node_records`. Disable this if CFG is disabled.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4263?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4263** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4263?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#4262** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4262?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

rzvxa reviewed (commented):
Wouldn't it be better if we do this at the call site? We might want to use these for other stuff. We can wrap both the `record_ast_node` and `retrive_recorded_ast_node` in `control_flow!` macros.
And since these 2 wouldn't get called without CFG we can remove the guards around `record_ast_node`.

overlookmotel committed:
perf(semantic): do not record ast nodes for cfg if cfg disabled (#4263)

Control flow graph builder records AST node IDs in a temp structure `ast_node_records`. Disable this if CFG is disabled.

overlookmotel commented:
> Wouldn't it be better if we do this at the call site? We might want to use these for other stuff. We can wrap both the `record_ast_node` and `retrive_recorded_ast_node` in `control_flow!` macros.

Yes, we could do that. But I didn't see the point of adding `control_flow!` at the call sites, because ultimately the macro does the same thing as the check I've put in `record_ast_nodes` etc.

Although... the macro would move the `if self.cfg.is_some()` check inline within `visit_*` functions. So it might make it faster when CFG is disabled (e.g. in transformer), as it'd remove a function call (if compiler doesn't inline these functions anyway - both a quite trivial).

We could alternatively try to get same optimization by marking these functions `#[inline]`. I'll try that first.

> And since these 2 wouldn't get called without CFG we can remove the guards around `record_ast_node`.

Yes, you're right. That guard could be removed.
